pczt/src/lib.rs

```diff
@@ -1 +1,129 @@
+//!   - Because we aren't storing the partial transaction in network format, we need to
+//!     carefully define equality for PCZTs.
+//!     - If we say "pczt.global must be identical" then:
+//!       - If we add spends or outputs in series, we should always update bsk when adding
+//!         spends or outputs, even if rcv is present.
+//!       - If we add spends or outputs in parallel and then combine, we must _never_ update
+//!         bsk, and then update it when we prepare for signature creation.
+//!       We can't control which happens, ergo we need an IO Finalizer step.
+//!     - Once every spend and output has its zkproof field set, PCZT equality MUST include
+//!       the SpendDescription and OutputDescription contents being identical.
+//!       - In practice enforced by creating a TransactionData / CMutableTransaction from
+//!         the PCZT, with spendAuthSigs and bindingSig empty, and then enforcing equality.
+//!       - This is equivalent to BIP 147's equality definition (the partial transactions
+//!         must be identical).
+//! - Spend Finalizer (anyone can execute)
+//!   - Combines partial transparent signatures into `script_sig`s.
+//! - Transaction Extractor (anyone can execute)
+//!   - Creates bindingSig and extracts the final transaction.
 
+use serde::{Deserialize, Serialize};
+
+pub mod roles;
+
+mod common;
+mod orchard;
+mod sapling;
+mod transparent;
+
+const MAGIC_BYTES: &[u8] = b"PCZT";
+const PCZT_VERSION_1: u32 = 1;
+
+#[cfg(feature = "zcp-builder")]
+const SAPLING_TX_VERSION: u32 = 4;
+const V5_TX_VERSION: u32 = 5;
+const V5_VERSION_GROUP_ID: u32 = 0x26A7270A;
+
+/// A partially-created Zcash transaction.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Pczt {
+    /// Global fields that are relevant to the transaction as a whole.
+    global: common::Global,
+
+    //
+    // Protocol-specific fields.
+    //
+    // Unlike the `TransactionData` type in `zcash_primitives`, these are not optional.
+    // This is because a PCZT does not always contain a semantically-valid transaction,
+    // and there may be phases where we need to store protocol-specific metadata before
+    // it has been determined whether there are protocol-specific inputs or outputs.
+    //
+    transparent: transparent::Bundle,
+    sapling: sapling::Bundle,
+    orchard: orchard::Bundle,
+}
+
+impl Pczt {
+    /// Parses a PCZT from its encoding.
+    pub fn parse(bytes: &[u8]) -> Result<Self, ParseError> {
+        if bytes.len() < 8 {
+            return Err(ParseError::TooShort);
+        }
+        if &bytes[..4] != MAGIC_BYTES {
+            return Err(ParseError::NotPczt);
+        }
+        let version = u32::from_le_bytes(bytes[4..8].try_into().unwrap());
+        if version != PCZT_VERSION_1 {
+            return Err(ParseError::UnknownVersion(version));
+        }
+
+        // This is a v1 PCZT.
+        postcard::from_bytes(&bytes[8..]).map_err(ParseError::Postcard)
```

Comment:
File a ticket for a test that exercises error cases.

Reply:
I'm adding the latter to this PR.